src/aiosalesforce/auth/jwt_flow.py

```diff
@@ -0,0 +1,139 @@
+        payload = {
+            "iss": self.client_id,
+            "sub": unescape(self.subject),
+            "aud": f"https://{'test' if sandbox else 'login'}.salesforce.com",
+            "exp": f"{expiration:.0f}",
+        }
+        with open(self.priv_key_file, "rb") as file:
+            priv_key = file.read()
+
+        if self.priv_key_passphrase:
+            from cryptography.hazmat.backends import default_backend
+            from cryptography.hazmat.primitives import serialization
+
+            passphrase = self.priv_key_passphrase.encode("utf-8")
+            private_key = serialization.load_pem_private_key(
+                priv_key, password=passphrase, backend=default_backend()
+            )
+        else:
+            private_key = priv_key
+
+        assertion = jwt.encode(
+            payload, private_key, algorithm="RS256", headers={"alg": "RS256"}
+        )
+        token_data = {
+            "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
+            "assertion": assertion,
+        }
+
+        request = client.httpx_client.build_request(
+            "POST",
+            f"{client.base_url}/services/oauth2/token",
+            headers={
+                "Content-Type": "application/x-www-form-urlencoded",
+            },
+            data=token_data,
+        )
+        await client.event_bus.publish_event(
+            RequestEvent(
+                type="request",
+                request=request,
+            )
+        )
+        retry_context = client.retry_policy.create_context()
+        response = await retry_context.send_request_with_retries(
+            httpx_client=client.httpx_client,
+            event_bus=client.event_bus,
+            semaphore=client._semaphore,
+            request=request,
+        )
+        if not response.is_success:
+            try:
+                response_json = json_loads(response.content)
+                error_code = response_json["error"]
+                error_message = response_json["error_description"]
+            except Exception:  # pragma: no cover
+                error_code = None
+                error_message = response.text
+            raise AuthenticationError(
+                f"[{error_code}] {error_message}" if error_code else error_message,
+                response=response,
+                error_code=error_code,
+                error_message=error_message,
+            )
+        await client.event_bus.publish_event(
+            ResponseEvent(
+                type="response",
+                response=response,
+            )
+        )
+        if self.timeout is not None:
+            self._expiration_time = time.time() + self.timeout
+        return json_loads(response.content)["access_token"]
+
+    #
+    # @property
+    # def expired(self) -> bool:
+    #     super().expired
+    #     if self._expiration_time is None:  # pragma: no cover
+    #         return False
+    #     return self._expiration_time <= time.time()
```

Comment:
Is there a reason this is commented out? Looks good to me.

Reply:
I undertand that there are no refresh tokens, but the JWT itself still expires. This means you need to request a new token when that happens. This property helps avoiding having a request fail with 401 before a retry.

Reply:
The JWT does not expire unless you're referring to the public certificate asssociate with the pub/priv key pair? 

What is returned from Salesforce is an access token (session id) and that expires. But there's no way that I know of to determine when a session is going to expire ahead of time so you just have to handle it when it does.

Reply:
Don't you set expiration yourself here?

```python
        expiration = int(time.time()) + 300
        sandbox = "sandbox." in client.base_url.lower()
        payload = {
            "iss": self.client_id,
            "sub": unescape(self.subject),
            "aud": f"https://{'test' if sandbox else 'login'}.salesforce.com",
            "exp": f"{expiration:.0f}",
        }
```

Reply:
That is the expiration of the authentication claim itself to prevent it from being abused/reused if it gets intercepted somehow. Once the access token is received, that assertion is no longer used. It's not like after that time elapses the session is no longer valid.

Reply:
In that case, how is the `timeout` parameter your declare in `__init__` used?

Reply:
You're right, that's not needed. That's a copy/paste remnant from ClientCredentialsFlow which I used as a basis.

Reply:
But how do you get access to the session timeout from Salesforce? I've never seen a way to know how long or when a session will timeout.

Reply:
I don't think there is a way to get session timeout from the token (session id) because its opaque. However, when you configure your connected app, you can specify session duration. By default it's not set so an api call will have to fail and get a new token when it expires. If duration is set, however, you can set expiration time when you fetch the token - it's current time + duration. When it expires, client will automaticall refresh it without having to make a call and retry it. Look at client credentials flow - that's how I implemented it there. It's also in the aiosalesforce docs here: https://georgebv.github.io/aiosalesforce/documentation/authentication/#client-credentials-flow